README.md
KenyaHMIS Application Toolbox
=============================
A javaFX application designed to handle deployment of various project applications. 

This initial version still supports downloading and deployment of KenyaEMR with support for more applications in future releases

### Requirements
1. java 17 and above - bundled into the package. Copy into </opt/java/toolkit/>
2. openjfx 17.0.8 - bundled into the package. Copy into </opt/openjfx/>
3. maven

### Packaging
1. Clone the project into a directory of choice
2. Navigate to the project's root directory
3. If you dont have JDK 17, download it from https://download.oracle.com/java/17/latest/jdk-17_linux-x64_bin.tar.gz and extract to </opt/java/toolkit/>
4. If you dont have JavaFX 17.0.8, download it from https://download2.gluonhq.com/openjfx/17.0.8/openjfx-17.0.8_linux-x64_bin-sdk.zip and extract to </opt/openjfx/>
5. If you dont have Maven, Install maven with <sudo apt update && sudo apt install maven>
6. Build using maven by running <./mvnw clean package> in the terminal
7. Navigate to the KenyaHMISToolKit directory
8. There is the kenyahmistoolkit.jar file. This needs to be uploaded for the release.
8. Run the packagedeb.sh <sudo ./packagedeb.sh> to build the .deb package. This can be used to deploy the toolkit on debian/ubuntu systems 
    NB: this gives you the kenyahmistoolkit-x.x.x.deb for the release. This needs to be uploaded for the release.
9. The build process also creates the following zipped files in the KenyaHMISToolKit directory:
        |-- openmrs-backup-tools.zip
        |-- rollback-tools.zip
        |-- scripts.zip
10. These zip files need to be uploaded during the release

This process uses maven shade plugin that provides a fat jar file that can be deployed. 

### RELEASE
1. KenyaEMR is normally released at: https://github.com/palladiumkenya/kenyahmis-releases/releases
    NB: This is where the toolkit looks for new KenyaEMR releases
    NB: Always tag the release with the version number of KenyaEMR e.g 18.6.2
        Folder structure of the release e.g for version 18.6.2:
            |--  KenyaEMR_18.6.2.zip
            |--  Kenyaemr.Release.Notes.v.18.6.2.pdf 

2. Toolkit is normally released at: https://github.com/palladiumkenya/kenyaemr-autoupdate/releases
    NB: This is where the toolkit looks for new KenyaEMR Toolkit releases
    NB: Always tag the release with the version number of Toolkit e.g 1.2.0
            Folder structure of the release e.g for version 1.2.0:
            |--  kenyahmistoolkit-1.2.0.deb  <-- debian package for ubuntu/debian based systems
            |--  kenyahmistoolkit.jar  <-- the latest jar file
            |--  openmrs-backup-tools.zip
            |--  rollback-tools.zip
            |--  scripts.zip

### Deployment

#### Required directory structure
The app is configured to use /opt/kehmisApplicationToolbox/Downloads directory for its operations. Create the directories using:

1. sudo mkdir /opt/kehmisApplicationToolbox/
2. sudo mkdir /opt/kehmisApplicationToolbox/Downloads
3. sudo chmod 777 -R /opt/kehmisApplicationToolbox/

You can launch the jar file using the below command. 

    java --module-path /opt/openjfx/lib --add-modules ALL-MODULE-PATH -jar /usr/share/kenyahmistoolkit/kenyahmistoolkit.jar

If your default java is not 17, please use the fully qualified path for java i.e. 

    /opt/java/toolkit/bin/java --module-path /opt/openjfx/lib --add-modules ALL-MODULE-PATH -jar /usr/share/kenyahmistoolkit/kenyahmistoolkit.jar

Or 
    Just use the desktop shortcut

Useful resources
- https://maven.apache.org/plugins/maven-shade-plugin/
- https://stackoverflow.com/questions/52130548/how-do-i-fix-javafx-runtime-components-are-missing
- https://openjfx.io/openjfx-docs/

#### References
```
REF: JavaFX - https://gluonhq.com/products/javafx/
     JavaFX - https://download2.gluonhq.com/openjfx/17.0.8/openjfx-17.0.8_linux-x64_bin-sdk.zip
REF: JDK 17 - https://download.oracle.com/java/17/latest/jdk-17_linux-x64_bin.tar.gz
REF: KenyaEMR Release - https://github.com/palladiumkenya/kenyahmis-releases/releases/latest
REF: kenyahmistoolkit Release - https://github.com/palladiumkenya/kenyaemr-autoupdate/releases/latest
```



KenyaHMISToolKit/changelog
kenyahmistoolkit (1.0.2) trusty; urgency=low
  * Rebuild
 -- Patrick Waweru <patrick.waweru@thepalladiumgroup.com>  Wed, 04 Oct 2023 16:05:23 +0300


KenyaHMISToolKit/packagedeb.sh
#!/bin/sh
 
PACKAGE_NAME="kenyahmistoolkit"
PACKAGE_VERSION="1.0.2"
SOURCE_DIR=$PWD
TEMP_DIR="/tmp/kenyahmistoolkit"
 
mkdir -p $TEMP_DIR/debian/DEBIAN
mkdir -p $TEMP_DIR/debian/lib
mkdir -p $TEMP_DIR/debian/usr/games
mkdir -p $TEMP_DIR/debian/usr/share/applications
mkdir -p $TEMP_DIR/debian/usr/share/$PACKAGE_NAME
mkdir -p $TEMP_DIR/debian/usr/share/doc/$PACKAGE_NAME
mkdir -p $TEMP_DIR/debian/usr/share/common-licenses/$PACKAGE_NAME

mkdir -p $TEMP_DIR/debian/opt/kehmisApplicationToolbox/Downloads/Scripts
mkdir -p $TEMP_DIR/debian/opt/kehmisApplicationToolbox/rollback

### Download and Inject the JDK
rm -rf $TEMP_DIR/debian/opt/java/toolkit/
mkdir -p $TEMP_DIR/debian/opt/java/toolkit/

if [ -e "jdk-17_linux-x64_bin.tar.gz" ]; then
    # If the jdk exists, just extract it
    echo "JDK file exists. Now extracting ..."
    # move file to destination
    mv jdk-17_linux-x64_bin.tar.gz $TEMP_DIR/debian/opt/java/toolkit/
    # tar -xvf jdk-17_linux-x64_bin.tar.gz -C $TEMP_DIR/debian/opt/java/toolkit/
    # extract it
    cd $TEMP_DIR/debian/opt/java/toolkit/ && tar -xvf jdk-17_linux-x64_bin.tar.gz --strip-components=1 --wildcards '*/' && cd $SOURCE_DIR
    rm -rf $TEMP_DIR/debian/opt/java/toolkit/jdk-17_linux-x64_bin.tar.gz
    chmod -R 777 $TEMP_DIR/debian/opt/
else
    # if the jdk doesnt exist, we attempt to download it
    wget -O jdk-17_linux-x64_bin.tar.gz https://download.oracle.com/java/17/latest/jdk-17_linux-x64_bin.tar.gz

    if [ $? -eq 0 ] && [ -e "jdk-17_linux-x64_bin.tar.gz" ]; then
        echo "JDK Download successful. Now extracting ..."
        # tar -xvf jdk-17_linux-x64_bin.tar.gz -C $TEMP_DIR/debian/opt/java/toolkit/
        # move file to destination
        mv jdk-17_linux-x64_bin.tar.gz $TEMP_DIR/debian/opt/java/toolkit/
        # extract it
        cd $TEMP_DIR/debian/opt/java/toolkit/ && tar -xvf jdk-17_linux-x64_bin.tar.gz --strip-components=1 --wildcards '*/' && cd $SOURCE_DIR
        rm -rf $TEMP_DIR/debian/opt/java/toolkit/jdk-17_linux-x64_bin.tar.gz
        chmod -R 777 $TEMP_DIR/debian/opt/
    else
        echo "JDK Download failed. No need to proceed"
        # Exit with code 11 -- Failed to install JDK
        exit 11
    fi
fi

echo "SUCCESS: JDK Setup Complete"

### Download and Inject the JavaFX SDK
rm -rf $TEMP_DIR/debian/opt/openjfx/
mkdir -p $TEMP_DIR/debian/opt/openjfx/

if [ -e "openjfx-17.0.8_linux-x64_bin-sdk.zip" ]; then
    # If the JavaFX SDK exists, just extract it
    echo "JavaFX SDK file exists. Now extracting ..."
    # move file to destination
    # mv openjfx-17.0.8_linux-x64_bin-sdk.zip $TEMP_DIR/debian/opt/openjfx/
    # unzip openjfx-17.0.8_linux-x64_bin-sdk.zip -d $TEMP_DIR/debian/opt/openjfx/
    mkdir tmp
    unzip openjfx-17.0.8_linux-x64_bin-sdk.zip -d tmp/
    mv tmp/*/* $TEMP_DIR/debian/opt/openjfx/
    rm -rf tmp
    chmod -R 777 $TEMP_DIR/debian/opt/
else
    # if the JavaFX SDK doesnt exist, we attempt to download it
    wget -O openjfx-17.0.8_linux-x64_bin-sdk.zip https://download2.gluonhq.com/openjfx/17.0.8/openjfx-17.0.8_linux-x64_bin-sdk.zip

    if [ $? -eq 0 ] && [ -e "openjfx-17.0.8_linux-x64_bin-sdk.zip" ]; then
        echo "JavaFX SDK Download successful. Now extracting ..."
        # unzip openjfx-17.0.8_linux-x64_bin-sdk.zip -d $TEMP_DIR/debian/opt/openjfx/
        mkdir tmp
        unzip openjfx-17.0.8_linux-x64_bin-sdk.zip -d tmp/
        mv tmp/*/* $TEMP_DIR/debian/opt/openjfx/
        rm -rf tmp
        rm -rf openjfx-17.0.8_linux-x64_bin-sdk.zip
        chmod -R 777 $TEMP_DIR/debian/opt/
    else
        echo "JavaFX SDK Download failed. No need to proceed"
        # Exit with code 12 -- Failed to install java FX
        exit 12
    fi
fi

echo "SUCCESS: JavaFX Setup Complete"
 
echo "Package: $PACKAGE_NAME" > $TEMP_DIR/debian/DEBIAN/control
echo "Version: $PACKAGE_VERSION" >> $TEMP_DIR/debian/DEBIAN/control
cat control >> $TEMP_DIR/debian/DEBIAN/control
cat preinst >> $TEMP_DIR/debian/DEBIAN/preinst
cat postinst >> $TEMP_DIR/debian/DEBIAN/postinst
 
cp *.desktop $TEMP_DIR/debian/usr/share/applications/
cp copyright $TEMP_DIR/debian/usr/share/common-licenses/$PACKAGE_NAME/ # results in no copyright warning
cp copyright $TEMP_DIR/debian/usr/share/doc/$PACKAGE_NAME/ # results in obsolete location warning

cp -rR Scripts/Scripts/* $TEMP_DIR/debian/opt/kehmisApplicationToolbox/Downloads/Scripts/

chmod 0644 $TEMP_DIR/debian/usr/share/doc/$PACKAGE_NAME/
chmod 0644 $TEMP_DIR/debian/usr/share/common-licenses/$PACKAGE_NAME/
chmod 0775 $TEMP_DIR/debian/DEBIAN/preinst
chmod 0775 $TEMP_DIR/debian/DEBIAN/postinst
chmod -R 0775 $TEMP_DIR/debian/DEBIAN
chmod -R 0775 $TEMP_DIR/debian/opt/kehmisApplicationToolbox
 
cp *.jar $TEMP_DIR/debian/usr/share/$PACKAGE_NAME/
cp $PACKAGE_NAME $TEMP_DIR/debian/usr/games/

#cp -r toolkitdependecies $TEMP_DIR/debian/usr/share/$PACKAGE_NAME/
 
echo "$PACKAGE_NAME ($PACKAGE_VERSION) trusty; urgency=low" > changelog
echo "  * Rebuild" >> changelog
echo " -- Patrick Waweru <patrick.waweru@thepalladiumgroup.com>  `date -R`" >> changelog
gzip -9c changelog > $TEMP_DIR/debian/usr/share/doc/$PACKAGE_NAME/changelog.gz
 
cp *.png $TEMP_DIR/debian/usr/share/$PACKAGE_NAME/
chmod 0644 $TEMP_DIR/debian/usr/share/$PACKAGE_NAME/*.png
 
PACKAGE_SIZE=`du -bs $TEMP_DIR/debian | cut -f 1`
PACKAGE_SIZE=$((PACKAGE_SIZE/1024))
echo "Installed-Size: $PACKAGE_SIZE" >> $TEMP_DIR/debian/DEBIAN/control
 
chown -R root $TEMP_DIR/debian/
chgrp -R root $TEMP_DIR/debian/
 
cd $TEMP_DIR/
dpkg --build debian
mv debian.deb $SOURCE_DIR/$PACKAGE_NAME-$PACKAGE_VERSION.deb
rm -rf $TEMP_DIR/debian


